envoy/config/config_validator.h

```diff
@@ -0,0 +1,77 @@
+#pragma once
+
+#include <string>
+#include <vector>
+
+#include "envoy/common/optref.h"
+#include "envoy/config/subscription.h"
+#include "envoy/server/instance.h"
+
+#include "source/common/protobuf/protobuf.h"
+
+namespace Envoy {
+namespace Config {
+
+/**
+ * The configuration validator interface. One can implement such validator
+ * to add custom constraints to the fetched config, and reject a config
+ * which violates these constraints.
+ * The validators will be extensions that can be dynamically configured.
+ *
+ * A validator example: a validator that prevents removing all Clusters
+ * from an Envoy (which may be caused by a bug in the config plane, and not
+ * intentional change).
+ */
+class ConfigValidator {
+public:
+  virtual ~ConfigValidator() = default;
+
+  /**
+   * Validates a given set of resources matching a State-of-the-World update.
+   * @param server A server instance to fetch the state before applying the config.
+   * @param resources List of decoded resources that reflect the new state.
+   * @throw optional EnvoyException if the config should be rejected.
+   * @return whether the config is valid or not.
```

Comment:
Why do we have both a `false` return option and an `EnvoyException` option to reject?

Reply:
As we discussed offline, changed to only throw an exception.